app/src/main/java/com/github/fribourgsdp/radio/FirestoreDatabase.kt

```diff
@@ -1,31 +1,65 @@
 package com.github.fribourgsdp.radio
 
+import android.content.ContentValues
+import android.util.Log
 import com.google.android.gms.tasks.Task
+import com.google.android.gms.tasks.Tasks
+import com.google.firebase.firestore.DocumentReference
 import com.google.firebase.firestore.DocumentSnapshot
 import com.google.firebase.firestore.ktx.firestore
 import com.google.firebase.ktx.Firebase
 import com.google.firebase.firestore.EventListener
 import java.lang.Exception
 
+/**
+ * This class serves to make possible the dependency injection with mockito. We mock the
+ * reference and serve documents snapshot that are also mocked.
+ *
+ */
+open class FirestoreRef {
+    val db = Firebase.firestore
+    open fun getUserRef(userId : String) : DocumentReference{
+        return db.collection("user").document(userId)
+    }
+    open fun getSongRef(songId : String) : DocumentReference{
+        return db.collection("songs").document(songId)
+    }
+    open fun getPlaylistRef(playlistId : String) : DocumentReference{
+        return db.collection("playlists").document(playlistId)
+    }
+}
 /**
  *
  *This class represents a database. It communicates with the database(Firestore)
  * and translates the result in classes
  *
  * @constructor Creates a database linked to Firestore
  */
-class FirestoreDatabase : Database {
+class FirestoreDatabase(var refMake: FirestoreRef) : Database {
     private val db = Firebase.firestore
-
+    constructor():this(FirestoreRef())
 
     override fun getUser(userId : String): Task<User> {
-        return  db.collection("user").document(userId).get().continueWith { l ->
+        return  refMake.getUserRef(userId).get().continueWith { l ->
             val result = l.result
             if(result.exists()){
-                User(result["first"].toString())
+                val userId = result["userID"].toString()
+                val playlists = result["playlists"] !! as ArrayList<HashMap<String,String>>

```

Comment:
Have you considered the case where the result has null values for playlists and other fields, i.e. everywhere you use `!!` ?

Reply:
I have not and I know it is kind of a strong assumption. But the only function setting a user is setUser, where we set these exact fields. It is not retro compatible but we don't need retro compatility toward an early phase of our project imo.

Reply:
Ok fair enough !